File: windows-native/ClaudeSession.cs
// Detects the currently active Claude Code session by scanning
// %USERPROFILE%\.claude\projects for the most recently modified .jsonl
// transcript. Port of AgentCord/ClaudeSession.swift.
//
// The transcript schema is undocumented, so all parsing is defensive:
// malformed or unexpected lines are skipped, never fatal. Scans are driven by
// the presence controller's tick (the macOS FSEvents watcher becomes a plain
// re-scan here; the per-file cache keeps that cheap).

using System.Diagnostics;
using System.Text.Json;
using System.Text.RegularExpressions;

namespace AgentCord;

public sealed class ClaudeSession
{
    /// <summary>A transcript counts as active if modified within this window.</summary>
    public double ActiveWindowSeconds { get; set; } = 60;

    /// <summary>When summing the day's working time, a gap between two consecutive
    /// messages longer than this is treated as a break (idle), not work, so a
    /// session left open does not inflate the total. This also excludes the gaps
    /// between separate sessions.</summary>
    private const long ActiveGapToleranceMs = 5 * 60 * 1000;

    private readonly string _projectsDir;
    private readonly Dictionary<string, CacheEntry> _aggregateCache = [];
    private readonly Dictionary<string, string> _repoNameCache = [];

    public ClaudeSession()
    {
        var home = Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);
        _projectsDir = Path.Combine(home, ".claude", "projects");
    }

    /// <summary>Scan the transcript tree and return the active session, or null
    /// when none is active. Not thread-safe; call from one thread.</summary>
    public SessionInfo? Scan()
    {
        List<(string Path, DateTime Mtime)> files;
        try
        {
            files = Directory
                .EnumerateFiles(_projectsDir, "*.jsonl", SearchOption.AllDirectories)
                .Select(p => (Path: p, Mtime: File.GetLastWriteTimeUtc(p)))
                .ToList();
        }
        catch
        {
            return null;
        }
        if (files.Count == 0) return null;

        var newest = files.MaxBy(f => f.Mtime);
        if ((DateTime.UtcNow - newest.Mtime).TotalSeconds > ActiveWindowSeconds) return null;

        // The presence shows daily totals: tokens summed across every transcript
        // touched today, and an elapsed timer that reflects the combined working
        // time of all of today's sessions (idle gaps between sessions excluded).
        // "Today" is the local calendar day, so the totals reset at midnight.
        var dayStartMs = new DateTimeOffset(DateTime.Today).ToUnixTimeMilliseconds();

        long totalTokensToday = 0;
        long totalActiveMs = 0;
        var activeAgg = new DayAggregate();
        foreach (var file in files)
        {
            var agg = Aggregate(file.Path, file.Mtime, dayStartMs);
            totalTokensToday += agg.TokensToday;
            totalActiveMs += agg.ActiveMsToday;
            if (file.Path == newest.Path) activeAgg = agg;
        }

        // Drop cache entries for transcripts that no longer exist.
        var live = files.Select(f => f.Path).ToHashSet();
        foreach (var stale in _aggregateCache.Keys.Where(k => !live.Contains(k)).ToList())
            _aggregateCache.Remove(stale);

        return MakeSessionInfo(newest.Path, newest.Mtime, activeAgg, totalTokensToday, totalActiveMs);
    }

    // --- Parsing

    /// <summary>Per-transcript figures restricted to today, from one .jsonl.</summary>
    private sealed class DayAggregate
    {
        public string? Cwd;
        public string? Model;
        /// <summary>Timestamp (epoch ms) of the last message recorded today, used
        /// to extend the active session's working time up to "now".</summary>
        public long? LastTodayMs;
        /// <summary>Working time today: the sum of gaps between consecutive
        /// messages, counting only gaps short enough to be continuous work.</summary>
        public long ActiveMsToday;
        public long TokensToday;
    }

    private sealed record CacheEntry(DateTime Mtime, long DayStartMs, DayAggregate Aggregate);

    /// <summary>Parsing every transcript on each scan would be wasteful, so
    /// results are memoized per file. An entry is reused only while the file is
    /// unmodified and we are still on the same calendar day (the day boundary
    /// changes which lines count as "today").</summary>
    private DayAggregate Aggregate(string path, DateTime mtime, long dayStartMs)
    {
        if (_aggregateCache.TryGetValue(path, out var entry)
            && entry.Mtime == mtime && entry.DayStartMs == dayStartMs)
        {
            return entry.Aggregate;
        }

        var agg = new DayAggregate();
        long? prevTodayMs = null;
        try
        {
            foreach (var line in File.ReadLines(path))
            {
                var trimmed = line.Trim();
                if (trimmed.Length == 0) continue;

                JsonDocument doc;
                try { doc = JsonDocument.Parse(trimmed); }
                catch { continue; }

                using (doc)
                {
                    var root = doc.RootElement;
                    if (root.ValueKind != JsonValueKind.Object) continue;

                    if (agg.Cwd is null
                        && root.TryGetProperty("cwd", out var cwd)
                        && cwd.ValueKind == JsonValueKind.String
                        && cwd.GetString() is { Length: > 0 } c)
                    {
                        agg.Cwd = c;
                    }

                    long? lineMs = null;
                    if (root.TryGetProperty("timestamp", out var ts) && ts.ValueKind == JsonValueKind.String)
                        lineMs = EpochMsFromIso(ts.GetString());
                    var isToday = (lineMs ?? long.MinValue) >= dayStartMs;

                    if (root.TryGetProperty("message", out var message) && message.ValueKind == JsonValueKind.Object)
                    {
                        if (message.TryGetProperty("model", out var model)
                            && model.ValueKind == JsonValueKind.String
                            && model.GetString() is { Length: > 0 } m
                            && m != "<synthetic>")
                        {
                            agg.Model = m;
                        }
                        if (isToday && message.TryGetProperty("usage", out var usage) && usage.ValueKind == JsonValueKind.Object)
                        {
                            agg.TokensToday += IntProp(usage, "input_tokens") + IntProp(usage, "output_tokens");
                        }
                    }

                    if (isToday && lineMs is long ms)
                    {
                        // Add the gap from the previous message only if it is short
                        // enough to count as continuous work (idle breaks excluded).
                        if (prevTodayMs is long prev)
                        {
                            var delta = ms - prev;
                            if (delta > 0 && delta <= ActiveGapToleranceMs) agg.ActiveMsToday += delta;
                        }
                        prevTodayMs = ms;
                        agg.LastTodayMs = ms;
                    }
                }
            }
        }
        catch
        {
            // File vanished or unreadable mid-scan; keep whatever we parsed.
        }

        _aggregateCache[path] = new CacheEntry(mtime, dayStartMs, agg);
        return agg;
    }

    private static long IntProp(JsonElement obj, string name) =>
        obj.TryGetProperty(name, out var v) && v.ValueKind == JsonValueKind.Number && v.TryGetInt64(out var n) ? n : 0;

    private SessionInfo MakeSessionInfo(
        string newestPath, DateTime newestMtime, DayAggregate active, long totalTokensToday, long totalActiveMs)
    {
        var projectName = DeriveProjectName(Path.GetFileName(Path.GetDirectoryName(newestPath)) ?? "");
        if (active.Cwd is not null) projectName = RepoName(active.Cwd);

        // totalActiveMs covers work up to each session's last logged message.
        // The active session is ongoing, so extend it from its last message to
        // "now" (while that gap stays within the work tolerance). Backdating
        // start by the total makes Discord's elapsed timer show the combined
        // working time of all of today's sessions.
        var nowMs = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
        var elapsedMs = totalActiveMs;
        if (active.LastTodayMs is long last)
        {
            var tail = nowMs - last;
            if (tail > 0 && tail <= ActiveGapToleranceMs) elapsedMs += tail;
        }

        return new SessionInfo
        {
            ProjectName = projectName.Length == 0 ? "Claude Code" : projectName,
            Model = active.Model is null ? null : PrettyModel(active.Model),
            StartEpochMs = nowMs - elapsedMs,
            TotalTokens = totalTokensToday,
            LastModifiedMs = new DateTimeOffset(newestMtime).ToUnixTimeMilliseconds(),
        };
    }

    /// <summary>Claude Code encodes the project's cwd into the directory name by
    /// replacing path separators with hyphens. As a fallback (when no cwd field
    /// is present) we take the trailing segment.</summary>
    private static string DeriveProjectName(string dir)
    {
        var parts = dir.Split('-', StringSplitOptions.RemoveEmptyEntries);
        return parts.Length > 0 ? parts[^1] : dir;
    }

    /// <summary>Resolve the repository name for a working directory. Prefers the
    /// git remote (so a worktree like ".../agentcord/abuja" still reports
    /// "agentcord"), then the git toplevel, then the directory name.</summary>
    private string RepoName(string cwd)
    {
        if (_repoNameCache.TryGetValue(cwd, out var cached)) return cached;

        var name = Path.GetFileName(cwd.TrimEnd('\\', '/'));
        if (string.IsNullOrEmpty(name)) name = cwd;

        if (RunGit(["-C", cwd, "config", "--get", "remote.origin.url"]) is { } remote)
        {
            var baseName = remote.Split('/', '\\')[^1];
            if (baseName.EndsWith(".git", StringComparison.OrdinalIgnoreCase)) baseName = baseName[..^4];
            if (baseName.Length > 0) name = baseName;
        }
        else if (RunGit(["-C", cwd, "rev-parse", "--show-toplevel"]) is { } top)
        {
            var baseName = Path.GetFileName(top.TrimEnd('\\', '/'));
            if (!string.IsNullOrEmpty(baseName)) name = baseName;
        }

        _repoNameCache[cwd] = name;
        return name;
    }

    private static string? RunGit(string[] args)
    {
        try
        {
            var psi = new ProcessStartInfo("git")
            {
                RedirectStandardOutput = true,
                RedirectStandardError = true,
                UseShellExecute = false,
                CreateNoWindow = true, // don't flash a console from the tray app
            };
            foreach (var a in args) psi.ArgumentList.Add(a);

            using var process = Process.Start(psi);
            if (process is null) return null;
            var output = process.StandardOutput.ReadToEnd().Trim();
            if (!process.WaitForExit(5000))
            {
                process.Kill();
                return null;
            }
            return process.ExitCode == 0 && output.Length > 0 ? output : null;
        }
        catch
        {
            return null;
        }
    }

    // --- Static helpers

    public static long? EpochMsFromIso(string? s)
    {
        if (string.IsNullOrEmpty(s)) return null;
        return DateTimeOffset.TryParse(
            s, System.Globalization.CultureInfo.InvariantCulture,
            System.Globalization.DateTimeStyles.RoundtripKind, out var dto)
            ? dto.ToUnixTimeMilliseconds()
            : null;
    }

    /// <summary>Turn a raw model id such as "claude-opus-4-5-20260101" into "Opus 4.5".</summary>
    public static string PrettyModel(string raw)
    {
        var lower = raw.ToLowerInvariant();
        string family;
        if (lower.Contains("opus")) family = "Opus";
        else if (lower.Contains("sonnet")) family = "Sonnet";
        else if (lower.Contains("haiku")) family = "Haiku";
        else if (lower.Contains("fable")) family = "Fable";
        else return raw;

        var match = Regex.Match(raw, "[0-9]+([.-][0-9]+)?");
        return match.Success ? $"{family} {match.Value.Replace('-', '.')}" : family;
    }
}

File: macos/AgentCord/ClaudeSession.swift
//
//  ClaudeSession.swift
//  AgentCord
//
//  Detects the currently active Claude Code session by watching
//  ~/.claude/projects/ and parsing the most recently modified .jsonl
//  transcript. Tokens are summed across transcripts touched today (local
//  calendar day). Elapsed time is the summed working duration across
//  transcripts that touched the last 24 hours (idle gaps excluded), matching
//  Grok / Codex / Cursor. The transcript schema is undocumented, so all
//  parsing is defensive: malformed or unexpected lines are skipped, never fatal.
//

import Foundation
import Combine

final class ClaudeSession: ObservableObject {

    /// The current active session, or nil when none is active.
    @Published private(set) var current: SessionInfo?

    /// A transcript counts as active if it was modified within this window.
    var activeWindowSeconds: TimeInterval = 60

    private let projectsURL: URL
    private let queue = DispatchQueue(label: "com.agentcord.session.scan", qos: .utility)
    private var eventStream: FSEventStreamRef?
    private var timer: DispatchSourceTimer?
    private var scanWorkItem: DispatchWorkItem?
    /// Queue-only. Start/stop flip this so an in-flight walk cannot republish
    /// after the monitor has been torn down.
    private var monitoring = false
    private var lastFullScan = Date.distantPast
    private var lastNewestDate: Date?
    private static let fullScanInterval: TimeInterval = 30
    private static let scanCoalesce: TimeInterval = 0.35

    init() {
        let home = FileManager.default.homeDirectoryForCurrentUser
        projectsURL = home.appendingPathComponent(".claude/projects", isDirectory: true)
    }

    func start() {
        guard timer == nil else { return }
        startFSEvents()
        startTimer()
        queue.async { [weak self] in
            self?.monitoring = true
            self?.scan()
        }
    }

    func stop() {
        if let stream = eventStream {
            FSEventStreamStop(stream)
            FSEventStreamInvalidate(stream)
            FSEventStreamRelease(stream)
            eventStream = nil
        }
        timer?.cancel()
        timer = nil
        scanWorkItem?.cancel()
        scanWorkItem = nil
        queue.async { [weak self] in
            self?.monitoring = false
            self?.lastNewestDate = nil
            self?.publish(nil)
        }
    }

    // MARK: File system monitoring

    private func startFSEvents() {
        var context = FSEventStreamContext(
            version: 0,
            info: Unmanaged.passUnretained(self).toOpaque(),
            retain: nil,
            release: nil,
            copyDescription: nil
        )
        let callback: FSEventStreamCallback = { _, info, _, _, _, _ in
            guard let info else { return }
            let session = Unmanaged<ClaudeSession>.fromOpaque(info).takeUnretainedValue()
            session.requestScan()
        }
        let paths = [projectsURL.path] as CFArray
        // Directory-level, coalesced events. FileEvents+NoDefer would fire a
        // full walk on every transcript append while Claude is working.
        guard let stream = FSEventStreamCreate(
            kCFAllocatorDefault,
            callback,
            &context,
            paths,
            FSEventStreamEventId(kFSEventStreamEventIdSinceNow),
            1.0,
            FSEventStreamCreateFlags(kFSEventStreamCreateFlagNone)
        ) else { return }

        FSEventStreamSetDispatchQueue(stream, queue)
        FSEventStreamStart(stream)
        eventStream = stream
    }

    private func startTimer() {
        // Cheap idle expiry plus a slow safety walk. FSEvents drive updates
        // while a session is active; walking the whole tree every 5s is waste.
        let t = DispatchSource.makeTimerSource(queue: queue)
        t.schedule(deadline: .now() + 5, repeating: 5)
        t.setEventHandler { [weak self] in self?.tick() }
        t.resume()
        timer = t
    }

    private func requestScan() {
        scanWorkItem?.cancel()
        let work = DispatchWorkItem { [weak self] in self?.scan() }
        scanWorkItem = work
        queue.asyncAfter(deadline: .now() + Self.scanCoalesce, execute: work)
    }

    private func tick() {
        guard monitoring else { return }
        if let newest = lastNewestDate, Date().timeIntervalSince(newest) > activeWindowSeconds {
            lastNewestDate = nil
            publish(nil)
            return
        }
        if Date().timeIntervalSince(lastFullScan) >= Self.fullScanInterval {
            scan()
        }
    }

    // MARK: Scanning (runs on `queue`)

    private func scan() {
        guard monitoring else { return }
        lastFullScan = Date()
        let fm = FileManager.default
        guard let enumerator = fm.enumerator(
            at: projectsURL,
            includingPropertiesForKeys: [.contentModificationDateKey],
            options: [.skipsHiddenFiles]
        ) else {
            publish(nil)
            return
        }

        var files: [(url: URL, date: Date)] = []
        var newest: (url: URL, date: Date)?
        for case let url as URL in enumerator where url.pathExtension == "jsonl" {
            let date = (try? url.resourceValues(forKeys: [.contentModificationDateKey]).contentModificationDate) ?? .distantPast
            files.append((url, date))
            if newest == nil || date > newest!.date {
                newest = (url, date)
            }
        }

        guard let newest else {
            lastNewestDate = nil
            publish(nil)
            return
        }
        // A fresh mtime is only a hint to parse. Orca/Claude Code can append a
        // timestamp-less `bridge-session` heartbeat and bump mtime on a dead
        // workspace; activity must come from parseable event timestamps.
        lastNewestDate = newest.date
        if Date().timeIntervalSince(newest.date) > activeWindowSeconds {
            publish(nil)
            return
        }

        // Tokens stay on the local calendar day. Elapsed time is a rolling 24h
        // sum of working gaps, same window as Grok / Codex / Cursor.
        let nowMs = Int64(Date().timeIntervalSince1970 * 1000)
        let cutoffMs = nowMs - SessionDuration.lookbackMs
        let dayStartMs = Int64(Calendar.current.startOfDay(for: Date()).timeIntervalSince1970 * 1000)

        var totalTokensToday = 0
        var totalActiveMs: Int64 = 0
        var newestLast: Int64?
        var best: (url: URL, date: Date, agg: DayAggregate, activity: Date)?
        for file in files {
            let agg = aggregate(url: file.url, mtime: file.date, dayStartMs: dayStartMs)
            totalTokensToday += agg.tokensToday
            let (activeMs, lastMs) = SessionDuration.activeMs(
                stamps: agg.stampsMs, cutoffMs: cutoffMs, nowMs: nowMs)
            totalActiveMs += activeMs
            if let lastMs, newestLast == nil || lastMs > newestLast! {
                newestLast = lastMs
            }
            let activity = agg.lastEventMs.map { Date(timeIntervalSince1970: TimeInterval($0) / 1000) }
                ?? file.date
            if best == nil || activity > best!.activity {
                best = (file.url, file.date, agg, activity)
            }
        }

        // Drop cache entries for transcripts that no longer exist.
        let liveURLs = Set(files.map { $0.url })
        aggregateCache = aggregateCache.filter { liveURLs.contains($0.key) }

        guard let best else {
            publish(nil)
            return
        }
        if Date().timeIntervalSince(best.activity) > activeWindowSeconds {
            publish(nil)
            return
        }

        guard monitoring else { return }
        publish(makeSessionInfo(
            newest: (best.url, best.activity),
            active: best.agg,
            totalTokensToday: totalTokensToday,
            startEpochMs: SessionDuration.startMs(
                totalActiveMs: totalActiveMs, lastMs: newestLast, nowMs: nowMs)
        ))
    }

    private func publish(_ info: SessionInfo?) {
        DispatchQueue.main.async { [weak self] in
            guard let self else { return }
            if self.current != info { self.current = info }
        }
    }

    // MARK: Parsing

    /// Per-transcript figures extracted from one `.jsonl`.
    private struct DayAggregate {
        var cwd: String?
        var model: String?
        /// Newest parseable event timestamp in the transcript (any day), used
        /// for idle detection so a heartbeat that only bumps mtime stays idle.
        var lastEventMs: Int64?
        /// Event timestamps, used to sum working time inside the 24h window.
        var stampsMs: [Int64] = []
        var tokensToday = 0
    }

    private struct CacheEntry {
        var mtime: Date
        var dayStartMs: Int64
        var cursor = JSONLCursor()
        var aggregate = DayAggregate()
    }

    /// Parsing every transcript on each scan would be wasteful, so results are
    /// memoized per file. Unchanged files reuse the last aggregate; a growing
    /// file only parses newly appended JSONL lines. Keyed access is safe:
    /// scanning is serialized on `queue`.
    private var aggregateCache: [URL: CacheEntry] = [:]

    private func aggregate(url: URL, mtime: Date, dayStartMs: Int64) -> DayAggregate {
        var entry = aggregateCache[url] ?? CacheEntry(mtime: mtime, dayStartMs: dayStartMs)
        if entry.mtime == mtime, entry.dayStartMs == dayStartMs {
            return entry.aggregate
        }

        if entry.dayStartMs != dayStartMs {
            entry = CacheEntry(mtime: mtime, dayStartMs: dayStartMs)
        }

        let pulled = entry.cursor.pullLines(from: url)
        if pulled.didReset {
            entry.aggregate = DayAggregate()
        }

        for line in pulled.lines {
            consumeAggregateLine(line, into: &entry.aggregate, dayStartMs: dayStartMs)
        }

        entry.mtime = mtime
        entry.dayStartMs = dayStartMs
        aggregateCache[url] = entry
        return entry.aggregate
    }

    private func consumeAggregateLine(
        _ line: String,
        into agg: inout DayAggregate,
        dayStartMs: Int64
    ) {
        let trimmed = line.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty, let data = trimmed.data(using: .utf8) else { return }
        guard let obj = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any] else { return }

        if agg.cwd == nil, let c = obj["cwd"] as? String, !c.isEmpty {
            agg.cwd = c
        }

        var lineMs: Int64?
        if let ts = obj["timestamp"] as? String { lineMs = Self.epochMs(fromISO: ts) }
        if let ms = lineMs {
            agg.lastEventMs = max(agg.lastEventMs ?? ms, ms)
            if ms >= dayStartMs - SessionDuration.lookbackMs {
                agg.stampsMs.append(ms)
            }
        }
        let isToday = (lineMs ?? .min) >= dayStartMs

        if let message = obj["message"] as? [String: Any] {
            if let m = message["model"] as? String, !m.isEmpty, m != "<synthetic>" {
                agg.model = m
            }
            if isToday, let usage = message["usage"] as? [String: Any] {
                agg.tokensToday += (usage["input_tokens"] as? Int ?? 0)
                agg.tokensToday += (usage["output_tokens"] as? Int ?? 0)
            }
        }
    }

    private func makeSessionInfo(
        newest: (url: URL, date: Date),
        active: DayAggregate,
        totalTokensToday: Int,
        startEpochMs: Int64
    ) -> SessionInfo {
        var projectName = deriveProjectName(fromDirectory: newest.url.deletingLastPathComponent().lastPathComponent)
        if let cwd = active.cwd { projectName = repoName(forCwd: cwd) }

        return SessionInfo(
            projectName: projectName.isEmpty ? "Claude Code" : projectName,
            model: active.model.map(Self.prettyModel),
            startEpochMs: startEpochMs,
            totalTokens: totalTokensToday,
            lastModified: newest.date
        )
    }

    /// Claude Code encodes the project's cwd into the directory name by
    /// replacing path separators with hyphens. As a fallback (when no `cwd`
    /// field is present) we take the trailing segment.
    private func deriveProjectName(fromDirectory dir: String) -> String {
        let parts = dir.split(separator: "-").filter { !$0.isEmpty }
        return parts.last.map(String.init) ?? dir
    }

    private var repoNameCache: [String: String] = [:]

    /// Resolve the repository name for a working directory. Prefers the git
    /// remote (so a Conductor worktree like ".../agentcord/abuja" still reports
    /// "agentcord"), then the git toplevel, then the directory name.
    private func repoName(forCwd cwd: String) -> String {
        if let cached = repoNameCache[cwd] { return cached }

        var name = (cwd as NSString).lastPathComponent
        if let remote = runGit(["-C", cwd, "config", "--get", "remote.origin.url"]) {
            var base = (remote as NSString).lastPathComponent
            if base.hasSuffix(".git") { base = String(base.dropLast(4)) }
            if !base.isEmpty { name = base }
        } else if let top = runGit(["-C", cwd, "rev-parse", "--show-toplevel"]) {
            let base = (top as NSString).lastPathComponent
            if !base.isEmpty { name = base }
        }

        repoNameCache[cwd] = name
        return name
    }

    private func runGit(_ args: [String]) -> String? {
        let process = Process()
        process.executableURL = URL(fileURLWithPath: "/usr/bin/git")
        process.arguments = args
        let pipe = Pipe()
        process.standardOutput = pipe
        process.standardError = FileHandle.nullDevice
        do { try process.run() } catch { return nil }
        let data = pipe.fileHandleForReading.readDataToEndOfFile()
        process.waitUntilExit()
        guard process.terminationStatus == 0 else { return nil }
        let output = String(data: data, encoding: .utf8)?.trimmingCharacters(in: .whitespacesAndNewlines)
        return (output?.isEmpty == false) ? output : nil
    }

    // MARK: Static helpers

    private static let isoWithFraction: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return f
    }()

    private static let isoPlain: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime]
        return f
    }()

    static func epochMs(fromISO string: String) -> Int64? {
        if let date = isoWithFraction.date(from: string) ?? isoPlain.date(from: string) {
            return Int64(date.timeIntervalSince1970 * 1000)
        }
        return nil
    }

    /// Turn a raw model id such as "claude-opus-4-5-20260101" into "Opus 4.5".
    static func prettyModel(_ raw: String) -> String {
        let lower = raw.lowercased()
        let family: String
        if lower.contains("opus") { family = "Opus" }
        else if lower.contains("sonnet") { family = "Sonnet" }
        else if lower.contains("haiku") { family = "Haiku" }
        else if lower.contains("fable") { family = "Fable" }
        else { return raw }

        if let range = raw.range(of: "[0-9]+([.-][0-9]+)?", options: .regularExpression) {
            let version = raw[range].replacingOccurrences(of: "-", with: ".")
            return "\(family) \(version)"
        }
        return family
    }
}
